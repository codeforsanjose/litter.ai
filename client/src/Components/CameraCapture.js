import React from 'react';


function CameraCapture () {
    return (
        <div>
            <input
            type="file" 
            accept=".jpg, .jpeg, .img, .HEIC"
            id="cameraFileInput"
            capture="environment"
            />
            <img id="pictureFromCamera" />

            <script src="ImageCapture.js"></script>
        </div>
    )
}

export default CameraCapture;