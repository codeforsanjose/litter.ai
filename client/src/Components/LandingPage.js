import React from 'react';
import { Link } from 'react-router-dom';

function LandingPage () {
    return (
        <div>
            <h1>Welcome to Litter.ai</h1>
            <p>With just a click, sort your garbage and save the world</p>
            <div style={{ position: 'fixed', bottom: '20px', left: '50%', transform: 'translateX(-50%)', width: '200px' }}>
                <button>Learn More</button>
                <Link to="/capture">
                    <button>Capture Image</button>
                </Link>
            </div>
        </div>
    )
}

export default LandingPage;