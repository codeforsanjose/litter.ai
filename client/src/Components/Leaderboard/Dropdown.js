import React from 'react';
import Select from 'react-select';

const options = [
  { value: 'total', label: 'Total' },
  { value: 'paper', label: 'Paper' },
  { value: 'metal', label: 'Metal' },
  { value: 'cardboard', label: 'Cardboard' },
  { value: 'trash', label: 'Trash' },
  { value: 'glass', label: 'Glass' },
  { value: 'plastic', label: 'Plastic' },
  { value: 'compost', label: 'Compost' },
  { value: 'others', label: 'Others' },
  { value: 'unknown', label: 'Unknown' },
]

const reactSelectStyles = {
  // Table header text label
  singleValue: (baseStyles) => ({
    ...baseStyles,
    backgroundColor: '#076B49',
    color: '#FFF',
  }),
  // Table header text border and background
  control: (baseStyles) => ({
    ...baseStyles,
    backgroundColor: '#076B49',
    border: 'none',
    width: '100%',
  }),
  // Dropdown menu
  option: (baseStyles, state) => ({
    ...baseStyles,
    backgroundColor:
    state.isSelected
    ? '#076B49'
    : state.isFocused
    ? '#74CC67' : '#FFF',
    color: state.isSelected ? '#FFF' : '#000',
    '&:hover': {
      color: '#FFF',
    },
    ':active': {
      ...baseStyles[':active'],
      backgroundColor: '#0c9265',
    }
  }),
  // Down arrow
  dropdownIndicator: (baseStyles) => ({
    ...baseStyles,
    color: '#FFF',
    "@media only screen and (max-width: 767px)": {
      ...baseStyles["@media only screen and (max-width: 767px)"],
      width: '1.5rem'
    },
  }),
}

export default function Dropdown({ setCategory }) {
  const handleChange = (option) => {
    setCategory(option.value);
  }

  return (
    <Select
      options={options}
      defaultValue={options[0]}
      onChange={handleChange}
      isSearchable={false}
      styles={reactSelectStyles}
    />
  )
}