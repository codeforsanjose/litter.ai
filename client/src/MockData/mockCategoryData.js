export const categoryData = {
  plastic: {
    name: 'Plastic',
    category: 'Recycle',
    icon: '',
    alt: 'recycle icon',
    description: `Not all plastic is recyclable, even if you see the chasing arrows sign!

    If the Plastic sign has a number #1 or #2 inside of the chasing arrows, it is most likely recyclable.

    To be recycled, plastics should be clean of food/residue.
    `
  },
  compost: {
    name: 'Compost',
    icon: '',
    alt: 'compost icon',
    description: 'This is compost. Lorem ipsum dolor sit amet, consectetur adipiscing elit, sed do eiusmod tempor incididunt ut labore et dolore magna aliqua. Consequat ac felis donec et odio pellentesque diam volutpat.'
  },
  trash: {
    name: 'Trash',
    icon: '',
    alt: 'trash icon',
    description: 'This is trash. Lorem ipsum dolor sit amet, consectetur adipiscing elit, sed do eiusmod tempor incididunt ut labore et dolore magna aliqua. Consequat ac felis donec et odio pellentesque diam volutpat.'
  },
}