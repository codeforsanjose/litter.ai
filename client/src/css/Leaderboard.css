.lb-container {
  height: 80vh;
  width: 42rem;
  max-width: 100vw;
  display: flex;
  flex-direction: column;
  align-items: center;
  justify-content: center;
}

.lb-container .lb-user-stats {
  width: 80%;
  height: 5vh;
  display: flex;
  align-items: center;
  justify-content: space-around;
  border: 0.2rem var(--main-green) solid;
  border-radius: 5rem;
  margin-bottom: 2rem;
}

.lb-container .lb-table {
  width: 95%;
  height: 60vh;
  border: 0.15rem var(--main-green) solid;
  border-spacing: inherit;
  text-align: center;
}

.lb-container .lb-table tr:nth-of-type(even) {
  background: var(--lb-gray);
}

.lb-container .lb-table th + th {
  border-left: 0.1rem var(--white)solid;
}

.lb-container .lb-table td + td {
  border-left: 0.1rem var(--main-green)solid;
}

.lb-container .lb-table .lb-header {
  background: var(--main-green);
  color: var(--white);
}

.lb-container .lb-table .lb-header th:nth-child(1) {
  width: 15%;
}

.lb-container .lb-table .lb-header th:nth-child(2) {
  height: 2.5rem;
  width: 20rem;
  text-align: left;
  padding-left: 1rem;
}

.lb-container .lb-table .lb-header th:nth-child(3) {
  width: 25%;
}

.lb-container .lb-name {
  text-align: left;
  padding-left: 1rem;
}

@media only screen and (max-width: 767px) {
  .lb-container {
    grid-area: 5 / 1 / span 1 / span 1;
  }

  .lb-container .lb-table .lb-header {
    font-size: 0.7rem;
  }

  .lb-container .lb-table .lb-header th:nth-child(3) {
    width: 30%;
  }

  .lb-container .lb-table tbody {
    font-size: 0.8rem;
  }
}

@media only screen and (min-width: 768px) {
  .lb-container {
    grid-area: 5 / 2 / span 1 / span 1;
  }
}