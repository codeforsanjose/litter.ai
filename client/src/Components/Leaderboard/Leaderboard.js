import React, { useState, useEffect } from 'react';
import Dropdown from './Dropdown';
import '../../css/Leaderboard.css';

export default function Leaderboard() {
  const [leaderboardData, setLeaderboardData] = useState([]);
  const [dropdownSelection, setDropdownSelection] = useState('total');
  const [userRank, setUserRank] = useState(0);

  // Creates a data row for each user with their rank, username, and total uploads
  const renderTable = (user, index) => (
    <tr key={index}>
      <td>{index + 1}</td>
      <td className="lb-name">{user.username}</td>
      <td>{dropdownSelection === 'total' ? user.totalUploads : user.itemCount}</td>
    </tr>
  )

  // Sorts users by total uploads
  useEffect(() => {
    // Data for the total uploads is a different link than the categories
    const link = dropdownSelection === 'total' ? 'http://localhost:3001/leaderboard' : `http://localhost:3001/leaderboard/${dropdownSelection}`;

    const fetchData = async () => {
      try {
        const res = await fetch(link);
        const response = await res.json();
        setLeaderboardData(response.leaderboard);
        // setUserRank(response.rank);
      }
      catch(err) {
        console.error(err);
      }
    }
    fetchData();
  }, [dropdownSelection])


  return (
    <div className="lb-container">
      <div className="lb-user-stats">
        <h3>Your rank: 9999</h3>
        <h3>Total: 99999999</h3>
      </div>
      <table className="lb-table">
        <thead>
          <tr className="lb-header">
            <th scope="col">Rank</th>
            <th scope="col">Name</th>
            <th scope="col"><Dropdown setCategory={setDropdownSelection}/></th>
          </tr>
        </thead>
      <tbody>
        {leaderboardData.map((user, index) => (renderTable(user, index)))}
      </tbody>
      </table>
    </div>
  )
}