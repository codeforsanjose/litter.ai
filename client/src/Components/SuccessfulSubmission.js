import React, { useState } from 'react';
import { Link } from 'react-router-dom';
import RecyclingIcon from '@mui/icons-material/Recycling';
import DeleteIcon from '@mui/icons-material/Delete';
import EggIcon from '@mui/icons-material/Egg';
import '../css/CategoryDescription.css'

export default function SuccessfulSubmission ({ type }) {
  const [category] = useState(type.category);
  return (
    <div className="category-container">
      <h4>{type.category}</h4>
      {category === 'Recycle' && <RecyclingIcon data-testid="recycle-icon" className="category-icon" />}
      {category === 'Trash' && <DeleteIcon data-testid="trash-icon" className="category-icon" />}
      {category === 'Compost' && <EggIcon data-testid="compost-icon" className="category-icon" />}
      <h1>{type.name}</h1>
      <p>{type.description}</p>
      <div className="category-buttons">
        <Link to="/learn-more"><button>Learn More</button></Link>
        <Link to="/capture"><button>Capture another photo</button></Link>
        <Link to="/"><button className="button-home">Home</button></Link>
      </div>
    </div>
  );
};
