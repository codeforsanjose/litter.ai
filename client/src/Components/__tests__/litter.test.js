import React from 'react';
import LandingPage from '../LandingPage';
import Leaderboard from '../Leaderboard/Leaderboard.js';
import CameraCapture from '../CameraCapture';
import SuccessfulSubmission from '../SuccessfulSubmission';
import { categoryData } from '../../MockData/mockCategoryData.js';
import {
  mockTotalUploads,
  mockPlasticUploads,
  mockMetalUploads
} from '../../MockData/mockLeaderboardData.js';
import renderer from 'react-test-renderer';
import { render, screen, fireEvent, waitFor } from '@testing-library/react';
import { createMemoryHistory } from 'history';
import { MemoryRouter, Routes, Route } from 'react-router-dom';

describe('Home page', () => {
  test('Landing page matches the current snapshot', () => {
    const tree = renderer.create(<LandingPage />).JSON();
    expect(tree).toMatchSnapshot();
  })
});

describe('Leaderboard component', () => {
  // Function to simulate an API call
  const mockCall = (data) => {
    jest.spyOn(global, 'fetch').mockResolvedValue({
      json: jest.fn().mockResolvedValue(data)
    });
  }

  // Start each test with total top 10 API call; reset mock after each test
  beforeEach(() => { mockCall(mockTotalUploads) });
  afterEach(() => { jest.restoreAllMocks() })

  // Creates a snapshot
  test('Leaderboard matches the current snapshot', () => {
    const tree = renderer.create(<Leaderboard />).toJSON();
    expect(tree).toMatchSnapshot();
  });

  test('Changes category from total to compost when clicking on the dropdown', () => {
    render(<Leaderboard />);
    // Clicks on dropdown
    fireEvent.mouseDown(screen.getByText('Total'));
    // Changes dropdown to 'Compost'
    fireEvent.click(screen.getByText('Compost'));
    // Checking if 'Compost' is now the header on the dropdown
    expect(screen.getByText('Compost')).toBeInTheDocument();
  });

  test('Dropdown option text and background changes color when hovered', () => {
    render(<Leaderboard />);
    // Clicks on dropdown
    fireEvent.mouseDown(screen.getByText('Total'));
    // Hover over new category
    fireEvent.mouseEnter(screen.getByText('Cardboard'));
    // Checking if 'Cardboard' has the hovered color and background color
    const mouseOverCardboard = getComputedStyle(screen.getByText('Cardboard'));
    expect(mouseOverCardboard.color).toBe('rgb(0, 0, 0)');
    expect(mouseOverCardboard.backgroundColor).toBe('rgb(116, 204, 103)');
  });

  test('New data is rendered when a dropdown category is changed', async () => {
    render(<Leaderboard />);
    // Change category and data to metal
    mockCall(mockMetalUploads);
    fireEvent.mouseDown(screen.getByText('Total'));
    fireEvent.click(screen.getByText('Metal'))
    // Checks if a specific user is in the document
    await waitFor(() => { expect(screen.getByText('lucious_senger10')).toBeInTheDocument() });
    // Change category and data to plastic
    fireEvent.mouseDown(screen.getByText('Metal'));
    mockCall(mockPlasticUploads);
    fireEvent.click(screen.getByText('Plastic'));
    // Checks if top metal user is no longer there and if a top plastic user is now visible
    await waitFor(() => {
      expect(screen.queryByText('lucious_senger10')).not.toBeInTheDocument();
      expect(screen.getByText('alejandra31')).toBeInTheDocument();
    });
  });
});

describe('Successful submission page', () => {
  test('Successful submission page matches the current snapshot', () => {
    // Memory router is used to allow Link component to work properly
    const tree = renderer.create(
      <MemoryRouter>
        <SuccessfulSubmission type={categoryData.plastic} />
      </MemoryRouter>
    )
    expect(tree).toMatchSnapshot();
  });

  test('Shows the description information for plastic', () => {
    render(
      <MemoryRouter>
        <SuccessfulSubmission type={categoryData.plastic} />
      </MemoryRouter>
    )
    expect(screen.getByText('Plastic')).toBeInTheDocument();
    expect(screen.getByText('Recycle')).toBeInTheDocument();
    expect(screen.getByTestId('recycle-icon')).toBeInTheDocument();
  });

  test('Navigates to the home page when "Home" is clicked', async () => {
    const history = createMemoryHistory({ initialEntries: ['/success'] });
    render(
      <MemoryRouter history={history} initialEntries={['/success']}>
        <Routes>
          <Route path='/success' element={<SuccessfulSubmission type={categoryData.plastic} />} />
          <Route path='/' element={<LandingPage />} />
        </Routes>
      </MemoryRouter>
    )
    // Initial endpoint is /success
    await waitFor(() => { expect(history.location.pathname).toBe('/success') });
    // Clicks button to navigate to new end point, then pushes endpoint into the history
    fireEvent.click(screen.getByRole('button', { name: 'Home' }));
    // Page currently navigates to /capture but doesn't update the endpoint in the history
    history.push('/');
    // Checks if page navigated to the landing page and the welcome header is displayed
    await waitFor(() => {
      expect(history.location.pathname).toBe('/')
      expect(screen.getByText('Welcome to Litter.ai')).toBeInTheDocument();
    });
  });

  test('Navigates to the capture page when "Capture another photo" is clicked', async () => {
    const history = createMemoryHistory({ initialEntries: ['/success'] });
    render(
      <MemoryRouter history={history} initialEntries={['/success']}>
        <Routes>
          <Route path='/success' element={<SuccessfulSubmission type={categoryData.plastic} />} />
          <Route path='/capture' element={<CameraCapture />} />
        </Routes>
      </MemoryRouter>
    )
    // Initial endpoint is /success
    await waitFor(() => { expect(history.location.pathname).toBe('/success') });
    // Clicks button to navigate to new end point, then pushes endpoint into the history
    fireEvent.click(screen.getByRole('button', { name: 'Capture another photo' }));
    // Page currently navigates to /capture but doesn't update the endpoint in the history
    history.push('/capture');
    // Checks if page navigated to the /capture endpoint
    await waitFor(() => { expect(history.location.pathname).toBe('/capture') });
  });
});