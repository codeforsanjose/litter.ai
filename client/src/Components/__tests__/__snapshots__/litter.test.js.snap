// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`First snapshot test 1`] = `
<div>
  <h1>
    Welcome to Litter.ai
  </h1>
  <p>
    With just a click, sort your garbage and save the world
  </p>
</div>
`;

exports[`Leaderboard component Leaderboard matches the current snapshot 1`] = `
<div
  className="lb-container"
>
  <div
    className="lb-user-stats"
  >
    <h3>
      Your rank: 9999
    </h3>
    <h3>
      Total: 99999999
    </h3>
  </div>
  <table
    className="lb-table"
  >
    <thead>
      <tr
        className="lb-header"
      >
        <th
          scope="col"
        >
          Rank
        </th>
        <th
          scope="col"
        >
          Name
        </th>
        <th
          scope="col"
        >
          <div
            className=" css-b62m3t-container"
            onKeyDown={[Function]}
          >
            <span
              className="css-1f43avz-a11yText-A11yText"
              id="react-select-2-live-region"
            />
            <span
              aria-atomic="false"
              aria-live="polite"
              aria-relevant="additions text"
              className="css-1f43avz-a11yText-A11yText"
            />
            <div
              className=" css-jc6pza-control"
              onMouseDown={[Function]}
              onTouchEnd={[Function]}
            >
              <div
                className=" css-1fdsijx-ValueContainer"
              >
                <div
                  className=" css-ayhps2-singleValue"
                >
                  Total
                </div>
                <input
                  aria-autocomplete="list"
                  aria-expanded={false}
                  aria-haspopup={true}
                  aria-readonly={true}
                  className="css-mohuvp-dummyInput-DummyInput"
                  disabled={false}
                  id="react-select-2-input"
                  inputMode="none"
                  onBlur={[Function]}
                  onChange={[Function]}
                  onFocus={[Function]}
                  role="combobox"
                  tabIndex={0}
                  value=""
                />
              </div>
              <div
                className=" css-1hb7zxy-IndicatorsContainer"
              >
                <span
                  className=" css-1u9des2-indicatorSeparator"
                />
                <div
                  aria-hidden="true"
                  className=" css-1s7cnte-indicatorContainer"
                  onMouseDown={[Function]}
                  onTouchEnd={[Function]}
                >
                  <svg
                    aria-hidden="true"
                    className="css-tj5bde-Svg"
                    focusable="false"
                    height={20}
                    viewBox="0 0 20 20"
                    width={20}
                  >
                    <path
                      d="M4.516 7.548c0.436-0.446 1.043-0.481 1.576 0l3.908 3.747 3.908-3.747c0.533-0.481 1.141-0.446 1.574 0 0.436 0.445 0.408 1.197 0 1.615-0.406 0.418-4.695 4.502-4.695 4.502-0.217 0.223-0.502 0.335-0.787 0.335s-0.57-0.112-0.789-0.335c0 0-4.287-4.084-4.695-4.502s-0.436-1.17 0-1.615z"
                    />
                  </svg>
                </div>
              </div>
            </div>
          </div>
        </th>
      </tr>
    </thead>
    <tbody />
  </table>
</div>
`;

exports[`Successful submission page Successful submission page matches the current snapshot 1`] = `
<div
  className="category-container"
>
  <h4>
    Recycle
  </h4>
  <svg
    aria-hidden={true}
    className="MuiSvgIcon-root MuiSvgIcon-fontSizeMedium category-icon css-i4bv87-MuiSvgIcon-root"
    data-testid="recycle-icon"
    focusable="false"
    viewBox="0 0 24 24"
  >
    <path
      d="M5.77 7.15 7.2 4.78l1.03-1.71c.39-.65 1.33-.65 1.72 0l1.48 2.46-1.23 2.06-1 1.62-3.43-2.06zm15.95 5.82-1.6-2.66-3.46 2L18.87 16H20c.76 0 1.45-.43 1.79-1.11.14-.28.21-.58.21-.89 0-.36-.1-.71-.28-1.03zM16 21h1.5c.76 0 1.45-.43 1.79-1.11L20.74 17H16v-2l-4 4 4 4v-2zm-6-4H5.7l-.84 1.41c-.3.5-.32 1.12-.06 1.65.28.57.87.94 1.52.94H10v-4zm-3.88-2.65 1.73 1.04L6.48 9.9 1 11.27l1.7 1.02-.41.69c-.35.59-.38 1.31-.07 1.92l1.63 3.26 2.27-3.81zm10.9-9.21-1.3-2.17C15.35 2.37 14.7 2 14 2h-3.53l3.12 5.2-1.72 1.03 5.49 1.37 1.37-5.49-1.71 1.03z"
    />
  </svg>
  <h1>
    Plastic
  </h1>
  <p>
    Not all plastic is recyclable, even if you see the chasing arrows sign!

    If the Plastic sign has a number #1 or #2 inside of the chasing arrows, it is most likely recyclable.

    To be recycled, plastics should be clean of food/residue.
    
  </p>
  <div
    className="category-buttons"
  >
    <a
      href="/learn-more"
      onClick={[Function]}
    >
      <button>
        Learn More
      </button>
    </a>
    <a
      href="/capture"
      onClick={[Function]}
    >
      <button>
        Capture another photo
      </button>
    </a>
    <a
      href="/"
      onClick={[Function]}
    >
      <button
        className="button-home"
      >
        Home
      </button>
    </a>
  </div>
</div>
`;
